### ISSUE TITLE
Clang can't find device library when using FHS compliant path

### ISSUE THREAD

**Issue by JacekJagosz** [RadeonOpenCompute/ROCm-Device-Libs#81]
I have packaged rocm for Solus closely following @Mystro256's packages for Fedora, not using `/opt/rocm` but instead `GNUInstallDirs`. The paths are `/usr/lib64/cmake/AMDDeviceLibs/AMDDeviceLibsConfig.cmake
` and`/usr/lib64/amdgcn/bitcode/`, same as on Fedora. I used [this patch without modifications](https://src.fedoraproject.org/rpms/rocm-device-libs/blob/rawhide/f/0001-Use-FHS-compliant-install.patch). I assume this will affect Fedora as well.
All packages work so far, up to rocm-opencl and hip, but when building rocBLAS I am getting:
```
clang-13: error: cannot find ROCm device library. Provide its path via --rocm-path or --rocm-device-lib-path, or pass -nogpulib to build without ROCm device library.
```
This is happening no matter if using LLVM 13 or 14, when defining `HIP_DEVICE_LIB_PATH`/`--rocm-device-lib-path` or not. 
I have tested so many different flags, went through how LLVM is supposed to detect device libs, and at this point have no idea what else to try. Does anyone have any suggestion what else can I try?
Here are my [rocm-device-libs build script](https://dev.getsol.us/D13161#change-dIsnTU8Mrbn6), and here the [rocBLAS one that fails](https://github.com/JacekJagosz/rocm-Solus/blob/main/rocblas/package.yml).

**Comment by Mystro256** [comment]
I think the Debian guys have a better patch. I've been focusing on getting opencl working first, so there might be holes in the hip/mathlib related components with my patches.

Take at look at what Debian did, because they have more testing HW and people.

**Comment by JacekJagosz** [comment]
Thanks for reply, but meanwhile I actually found the cause: https://reviews.llvm.org/D119478
It seems AMD renamed renoir from gfx902 to gfx90c, but didn't add 90c to supported architectures in Libomptarget.
So to fix it this one-line patch for LLVM needs to be included. It is a pity it is not in LLVM 14, but fortunately my distribution will include this patch for LLVM.

**Comment by JacekJagosz** [comment]
Commented and closed the wrong issue, really sorry about that.
I followed your work instead of Debian, because you did it for newer ROCm and LLVM versions, and didn't support my APU so I couldn't even test it. But I will follow their work more closely from now on, thank you!

**Comment by JacekJagosz** [comment]
I think the problem lies in LLVM itself. It is very hard to diagnose. I [found the function it is failing at](https://github.com/llvm/llvm-project/blob/f1e7ecaa18a7b1c77fc95e4b83d3db02c4e2d211/clang/lib/Driver/ToolChains/AMDGPU.cpp#L362), but have no idea why `RocmInstallationDetector::detectDeviceLibrary()` is failing.

While looking I found another problem, with the directories it is searching in for HIP: https://github.com/llvm/llvm-project/blob/50f846d63443781a5c00ec7022c381c4f09f543d/mlir/lib/ExecutionEngine/CMakeLists.txt#L191
```
  if (NOT DEFINED ROCM_PATH)
    if (NOT DEFINED ENV{ROCM_PATH})
      set(ROCM_PATH "/opt/rocm" CACHE PATH "Path to which ROCm has been installed")
    else()
      set(ROCM_PATH $ENV{ROCM_PATH} CACHE PATH "Path to which ROCm has been installed")
    endif()
    set(HIP_PATH "${ROCM_PATH}/hip" CACHE PATH "Path to which HIP has been installed")
  endif()
  set(CMAKE_MODULE_PATH "${HIP_PATH}/cmake" ${CMAKE_MODULE_PATH})
  find_package(HIP)
  if (NOT HIP_FOUND)
    message(SEND_ERROR "Building mlir with ROCm support requires a working ROCm and HIP install")
  else()
    message(STATUS "ROCm HIP version: ${HIP_VERSION}")
  endif()

  # Locate HIP runtime library.
  find_library(ROCM_RUNTIME_LIBRARY amdhip64
               PATHS "${HIP_PATH}/lib")
```
With that logic it is not possible to find both `/usr/include/hip/` and `/usr/lib64/cmake/Modules/` with one `HIP_PATH`.
Should I report the second problem in some other bugtracker, like LLVM?

**Comment by JacekJagosz** [comment]
I have reported in on LLVM's bugtracker as it is probably the culprit.

**Comment by Mystro256** [comment]
This might be related to the following patch, which landed in the main branch of clang (15), but I requested backporting to clang 14 in fedora:

https://src.fedoraproject.org/rpms/clang/pull-request/156

As far as I know, it resolves an issue when installing hip into /usr. @cgmb recommended it, but I didn't look too much into the background/details.

**Comment by Mystro256** [comment]
Sorry I missed your comment in the hipamd issue, in which you said you already use this patch.

**Comment by cgmb** [comment]
> This might be related to the following patch, which landed in the main branch of clang (15), but I requested backporting to clang 14 in fedora: https://src.fedoraproject.org/rpms/clang/pull-request/156 As far as I know, it resolves an issue when installing hip into /usr. @cgmb recommended it, but I didn't look too much into the background/details.

[That patch](https://github.com/llvm/llvm-project/commit/cd597588217a69b413ad44ba11d650b576045e2b) has been backported in LLVM upstream as well. It's included in Clang 14.0.4 or later. However, I believe it's entirely unrelated to this problem.

**Comment by JacekJagosz** [comment]
I have managed to fix this problem. I looked at [Debian's build of rocBLAS](https://salsa.debian.org/rocm-team/hipblas/-/blob/master/debian/rules), and after defining the following variables that they do too:
``` bash
export HIP_PLATFORM=amd
export HIP_RUNTIME=rocclr
export HIP_COMPILER=clang
export HIP_CLANG_PATH=/usr/bin
export DEVICE_LIB_PATH=/usr/lib64/amdgcn/bitcode
export HIP_DEVICE_LIB_PATH=/usr/lib64/amdgcn/bitcode
export HSA_PATH=/usr/hsa
export LLVM_PATH=/usr
```
I am no longer running into this specific issue. I didn't know they were necessary, [build for Arch didn't need them](https://github.com/rocm-arch/rocm-arch/blob/master/rocblas/PKGBUILD), but I guess this is specifically a problem when using FHS paths.
Now I am running into other issues when compiling, but at least this is solved.
